### ISSUE TITLE
Separation of retrieve and download

### ISSUE THREAD

**Issue by jblarsen** [ecmwf/cdsapi#2]
If retrieve is called without a target filename the method will execute a retrieve request to CDS and poll it until the request is 'completed' or 'failed' and return a Result object after that. The download method on this object can then be called to actually download the data file.

But for retrieve requests which take a very long time to complete (> 1 day) this workflow may not be suitable (on all systems). I would therefore propose an option to have a slightly different workflow:

1. The retrieve request returns a Result object after the first ("robust") request without waiting for the state of the reply to be 'completed' or 'failed'.
2. A new instance method on the Result class named e.g. 'query_state', 'update', 'update_state' or something like this is added.

This would allow users more control over the retrieval and download process. In our system it is for example not optimal to have very long running processes. We can then split the processing up into one task which executes the retrieve request and puts the Result object information in our key value store. After that we can then regularly dispatch a task which queries the state of the CDS task and downloads the resulting dataset when ready. Just like you do internally in cdsapi now.

So in summary this suggestion will not change anything about how cdsapi works now. But it will give users more flexibility in integrating cdsapi in data processing pipelines. Please let me know what you think of the above. Please see PR below for details.

**Comment by rabernat** [comment]
This would be a perfect use case for asyncio: https://docs.python.org/3/library/asyncio.html

**Comment by j08lue** [comment]
> asyncio

Yes, that would be pretty sweet. 

First step, I guess, is to get the attention of the maintainers.

**Comment by fridgerator** [comment]
Its even possible to re-create the Result object at a later time, for instance in another process or in a celery worker.

```python
import cdsapi

c = cdsapi.Client(wait_until_complete=False)
r = c.retrieve(...)
request_id = r.reply['request_id'] # save this for later
```
At a later time:

```python
reply = dict(request_id=request_id) # request_id from above
new_client = cdsapi.Client(wait_until_complete=False)
result = cdsapi.api.Result(new_client, reply)
result.update()
reply = result.reply
if reply['state'] == 'completed':
    result.download('file.nc')
```

**Comment by jblarsen** [comment]
Yes, definitely is. Closing issue:-)
